pkg/plugin/resource/resource.go

```diff
@@ -0,0 +1,200 @@
+			for _, d := range descriptions {
+				ids = append(ids, d.ID)
+			}
+			return "", fmt.Errorf("Found multiple resources named %s: %v", name, ids)
+		}
+	}
+	numFound := len(resourceIDs)
+
+	for _, name := range orderedResourceNames {
+		if _, ok := resourceIDs[name]; ok {
+			continue
+		}
+
+		resource := resources[name]
+		properties, err := executeAsTemplate(json.RawMessage(*resource.config.Properties), struct{ Resources interface{} }{resourceIDs})
+		if err != nil {
+			return "", fmt.Errorf("Failed to get properties for %s: %s", name, err)
+		}
+
+		id, err := resource.plugin.Provision(instance.Spec{
+			Properties: &properties,
+			Tags:       map[string]string{resourceGroupTag: string(config.ID), resourceNameTag: name},
+		})
+		if err != nil {
+			return "", fmt.Errorf("Failed to provision %s: %s", name, err)
+		}
+		log.Infof("Provisioned %s with ID %s", name, *id)
+		resourceIDs[name] = struct{ instance.ID }{*id}
+	}
+
+	return fmt.Sprintf("Found %d resources and created %d ", numFound, len(resourceIDs)-numFound), nil
+}
+
+func executeAsTemplate(text json.RawMessage, data interface{}) (json.RawMessage, error) {
+	tmpl, err := template.New("").Option("missingkey=error").Parse(string(text))
+	if err != nil {
+		return nil, err
+	}
+
+	var b bytes.Buffer
+	if err = tmpl.Execute(&b, data); err != nil {
+		return nil, err
+	}
+
+	return b.Bytes(), nil
+}
+
+var resourceReferenceRegexp = regexp.MustCompile(`{{\s*\.Resources\.(\w+)`)
+
+func getResourceReferences(properties json.RawMessage) []string {
+
+	var references []string
+	// TODO: Use text/template.Template.Execute instead.
+	for _, submatches := range resourceReferenceRegexp.FindAllSubmatch(properties, -1) {
+		references = append(references, string(submatches[1]))
+	}
+	return references
+}
+
+func getProvisioningOrder(resources map[string]*resource) ([]string, error) {
+	g := graph.New(graph.Directed)
+
+	nodes := map[string]graph.Node{}
+	for name := range resources {
+		nodes[name] = g.MakeNode()
+		*nodes[name].Value = name
+	}
+
+	for name, resource := range resources {
+		to := nodes[name]
+		references := getResourceReferences(json.RawMessage(*resource.config.Properties))
+		for _, reference := range references {
+			from, ok := nodes[reference]
+			if !ok {
+				return nil, fmt.Errorf("Reference to undefined resource %s in %s", reference, name)
+			}
+			if err := g.MakeEdge(from, to); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	for _, connectedComponent := range g.StronglyConnectedComponents() {
+		if len(connectedComponent) > 1 {
+			var cycle []string
+			for _, node := range connectedComponent {
+				cycle = append(cycle, (*node.Value).(string))
+			}
+			sort.Strings(cycle)
+			return nil, fmt.Errorf("Cycle detected: %v", cycle)
+		}
+	}
+
+	var sorted []string
+	for _, node := range g.TopologicalSort() {
+		sorted = append(sorted, (*node.Value).(string))
+	}
+
+	return sorted, nil
+}
+
+func (p *plugin) FreeGroup(id group.ID) error {
```

Comment:
@nwt  - how about this...   let's not make the resource plugin implement the group SPI.   Rather, let's define a new SPI:
   1. Add a new package:  `pkg/spi/resource`
   2. Let's create a new `spi.go` and define something with what we for sure need for the resource plugin:
```go
type Plugin interface {
    Commit(spec *resource.Spec, pretend bool) (plan string, err error)
    Destroy(resource resource.ID) error
    DescribeResources() ([]resource.Description, error) 
}
```

In essence, the Resource plugin manages a collection of resources (which are created by the instance/<type> plugins).   It does things like analyze dependencies among this collection of resources in the spec JSON and calls the instance plugin for creation, etc.   It also provides the point of contact for end users to list, commit and destroy resources.